modules/core/DecodeError.scala

```diff
@@ -0,0 +1,174 @@
+/* -
+ * Case Classy [classy-core]
+ */
+
+package classy
+package core
+
+import scala.annotation.tailrec
+
+/** An error that occurred while decoding data
+  */
+sealed abstract class DecodeError extends Product with Serializable {
+
+  /** Create a conjunction between this error and another */
+  def &&(other: DecodeError): DecodeError = DecodeError.and(this, other)
+
+  /** Create a disjunction between this error and another */
+  def ||(other: DecodeError): DecodeError = DecodeError.or(this, other)
+
+  /** Indicate that this error occured while processing a value at `path` */
+  def atPath(path: String): DecodeError.AtPath = DecodeError.AtPath(path, this)
+
+  /** Indicate that this error occurred while processing a value at `index` */
+  def atIndex(index: Int): DecodeError.AtIndex = DecodeError.AtIndex(index, this)
+
+  /** Is this error an aggregation of other errors? */
+  def isAggregate: Boolean = false
+}
+
+object DecodeError extends DecodeErrorInstances {
+
+  /** A marker for decode errors that don't have any child errors
+    */
+  sealed trait LeafDecodeError extends DecodeError
```

Comment:
Should have `extends Product with Serializable` in all ADTs. Also if there is no reason for extension abstract classes will be better suited. Serializable in case this ever gets used inside a Spark lambda that is serialized and sent across the wire.